nixos/binfmt: Add support for using statically-linked QEMU

zhaofengli opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This PR adds support for using a statically-linked build of QEMU without any intermediate wrapper to the binfmt-misc integration in NixOS. It also enables the "F" (fix binary) flag, allowing chroot into a foreign root filesystem without first having to make the interpreter available inside the rootfs. This fixes #160300 so foreign chroots "just work."

To use the statically-linked QEMU, set `boot.binfmt.preferStaticEmulators = true;`.

Currently, `qemu-user-static` is built with musl and has a couple of hard-to-debug problems. When building `coreutils` for aarch64-linux on x86_64-linux with it (see #143060 for background), <s>all tests pass except for `test-free` where it segfaults</s> the tests become stuck. Other distros ship `qemu-user-static` with static glibc. To prevent introducing new breakages, this PR keeps the dynamically-linked QEMU as the default.

## How To Test

1. Run `nix-build -A nixosTests.systemd-binfmt`
	- The `chroot` test is added which makes use of the new `qemu-user-static`
2. Activate a config with this PR applied
3. Confirm that `nix-build --system aarch64-linux -A hello --check` works
4. Activate a config with `boot.binfmt.preferStaticEmulators = true;`
5. Open `/etc/nix/nix.conf` and observe that `extra-sandbox-paths` is now empty
	- Yay purity!
6. Confirm that `nix-build --system aarch64-linux -A hello --check` still works
	- Sadly `nix-build --system aarch64-linux -A coreutils --check` doesn't work now
7. Try chrooting into [ArchLinuxARM-aarch64-latest.tar.gz](http://os.archlinuxarm.org/os/ArchLinuxARM-aarch64-latest.tar.gz)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

eddyb commented:
Have there been any updates?

I ran into the need for `aarch64` containers (to try and reproduce some weird `ar` archive corruption/format mismatch), and [ended up rebasing this PR](https://github.com/eddyb/nixpkgs/commits/binfmt-qemu-static) (of note is `guestAgentSupport` that was added after this PR, and I also ended up adding a commit for `numaSupport = false;` - not sure why that's only a problem now?).

For the limited testing needs I have I can probably just keep rebasing if I need to update, and just disable the `binfmt` config once I no longer need it - but it would be great to land this functionality in nixpkgs.

SuperSandro2000 commented on pkgs/applications/virtualization/qemu/user-static.nix:
```diff
@@ -0,0 +1,47 @@
+  qemuTargets = map (system: (lib.systems.elaborate { inherit system; }).qemuArch + "-linux-user") qemuUserPlatforms;
+
+  static = (pkgsStatic.qemu.override {
+    smartcardSupport = false;
+    spiceSupport = false;
+    openGLSupport = false;
+    virglSupport = false;
+    vncSupport = false;
+    gtkSupport = false;
+    sdlSupport = false;
+    alsaSupport = false;
+    jackSupport = false;
+    pulseSupport = false;
+    smbdSupport = false;
+    seccompSupport = false;
+    uringSupport = false;
+    systemSupport = false;
+    ncursesSupport = false;
+    hostCpuTargets = qemuTargets;
```
Maybe we should combine this into a minimal option or package?

YisuiDenghua replied:
```suggestion
    hostCpuTargets = qemuTargets;
    numaSupport = false;
```

tartavull commented:
This is great work, how can we contribute to finalize it?

zhaofengli commented:
I guess we can first get the support for building QEMU statically as well as `qemu-user-static` in first. After that, we can iterate on it and finally replace the binfmt implementation to use `qemu-user-static` in another PR.

YisuiDenghua commented:
Error occurred when I was trying to build NixOS from `github:zhaofengli/nixpkgs/binfmt-qemu-static`. Adding `numaSupport = false` to `user-binfmt.nix` may solve the problem. 
![图片](https://user-images.githubusercontent.com/102890144/204480690-504c8a09-bde0-4f3f-9859-69868c44451c.png)

Detegr commented:
This looks great, I'd love to see this getting merged.

zhaofengli commented:
Hi all, glad to see this finally getting some attention. I'm currently busy with something else at the moment but will get back to this PR in the coming days.

Detegr commented:
Is there anything I can do to help getting this forward?

hh-vincit reviewed (commented):
Looks good to me.

Detegr commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -111,8 +117,14 @@ stdenv.mkDerivation rec {
     })
+    # Adds binfmt P support, activated when argv[0] ends with `-binfmt-P`.
+    (fetchpatch {
+      url = "https://salsa.debian.org/qemu-team/qemu/-/raw/master/debian/patches/linux-user-binfmt-P.diff";
```
This does not exist anymore, removed in [this commit](https://salsa.debian.org/qemu-team/qemu/-/commit/1f805a1389b80ecda82a07b4ec190a0352fa3eb0).

I don't understand all details here, could this implementation work without the patch as well?

Detegr replied:
[Raw link to the patch](https://salsa.debian.org/qemu-team/qemu/-/raw/7551fa04dd734788faa8d6ee6fec6616dff88331/debian/patches/linux-user-binfmt-P.diff) if we still need it.

zhaofengli replied:
Yes, it indeed can now and is independent of how the emulator is named. I'll remove this patch.

Detegr commented:
Needs a rebase and fixing the issue with the patch I mentioned.

Tested this briefly and it seemed to work with no problems. Are there any other issues why this should not be merged?

alyssais commented on pkgs/applications/virtualization/qemu/user-static.nix:
```diff
@@ -0,0 +1,51 @@
+    "x86_64-linux"
+  ];
+
+  qemuTargets = map (system: (lib.systems.elaborate { inherit system; }).qemuArch + "-linux-user") qemuUserPlatforms;
```
```suggestion
  qemuTargets = map (system: (lib.systems.elaborate { inherit system; }).qemuArch + "-user") qemuUserPlatforms;
```

Nothing Linux-specific here, as QEMU also supports user-mode emulation of other (BSD) operating systems.

alyssais replied:
This was marked resolved, but nothing seems to have happened?

zhaofengli replied:
Oops sorry, I thought I took care of everything but I forgot about this.

zhaofengli replied:
I tried and it didn't work. The targets must have the `-linux-user` or `-bsd-user` suffixes. We can add `-bsd-user` after testing in a separate PR.

alyssais replied:
Oh yes I'm sorry, I misunderstood what this code did.  My bad.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -145,7 +155,10 @@ stdenv.mkDerivation rec {
     "--cpu=${stdenv.hostPlatform.uname.processor}"
     (lib.enableFeature guestAgentSupport "guest-agent")
-  ] ++ lib.optional numaSupport "--enable-numa"
+  ] ++ lib.optional enableDocs "--enable-docs"
```
Let's use `lib.enableFeature` and friends for new options, like it done for `guestAgentSupport` above.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -111,8 +117,14 @@ stdenv.mkDerivation rec {
   ]
-  ++ lib.optional nixosTestRunner ./force-uid0-on-9p.patch;
+  ++ lib.optional nixosTestRunner ./force-uid0-on-9p.patch
+  ++ lib.optional stdenv.hostPlatform.isStatic ./aio-find-static-library.patch;
```
What would it take for this to be upstreamable?

zhaofengli replied:
This seems to be [already gone](https://github.com/qemu/qemu/blob/ac84b57b4d74606f7f83667a0606deef32b2049d/meson.build#L856) since https://github.com/qemu/qemu/commit/063d5119c18af76b9e3303856dd114b36a8e1a43 and should land in QEMU 8.1. I'll add a comment about this.

zhaofengli committed:
qemu: Support disabling TCG plugins

The plugin .so's do not build statically.

zhaofengli committed:
qemu: Skip doc-related dependencies when docs are disabled

zhaofengli committed:
qemu: Support disabling tools

Not needed for a minimal qemu-user-static build.

zhaofengli committed:
nixos/tests/systemd-binfmt: Add chroot test

zhaofengli committed:
nixos/binfmt: Add option to use static emulators when available

The fixBinary flag will be enabled if a static emulator is in use.

zhaofengli committed:
qemu-user-static: init

This is synonymous with the "qemu-user-static" packages available in
other distros.

zhaofengli committed:
qemu: Only link qemu-kvm if it exists

zhaofengli committed:
qemu: Support disabling blobs

Not needed for a minimal qemu-user-static build.

zhaofengli committed:
qemu: Fix static build

zhaofengli commented:
Ok, I rebased the PR and it's now ready for review again. The default config is unchanged (dynamically-linked QEMU) so it's unlikely to introduce new breakages. To use statically-linked QEMU, set `boot.binfmt.preferStaticEmulators = true;`.

For the default case with dynamically-linked QEMU, we unfortunately still need the wrapper despite preserve-argv0 support having been upstreamed to strip out the `LD_*` environment variables so they don't cause spurious prints like

```
ERROR: ld.so: object 'some-foreign.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
```

## How To Test

1. Run `nix-build -A nixosTests.systemd-binfmt`
	- The `chroot` test is added which makes use of the new `qemu-user-static`
2. Activate a config with this PR applied
3. Confirm that `nix-build --system aarch64-linux -A hello --check` works
4. Activate a config with `boot.binfmt.preferStaticEmulators = true;`
5. Open `/etc/nix/nix.conf` and observe that `extra-sandbox-paths` is now empty
	- Yay purity!
6. Confirm that `nix-build --system aarch64-linux -A hello --check` still works
	- Sadly `nix-build --system aarch64-linux -A coreutils --check` doesn't work now
7. Try chrooting into [ArchLinuxARM-aarch64-latest.tar.gz](http://os.archlinuxarm.org/os/ArchLinuxARM-aarch64-latest.tar.gz)

alyssais commented on lib/systems/default.nix:
```diff
@@ -213,6 +213,10 @@ rec {
           else throw "Don't know how to run ${final.config} executables.";
 
+        staticEmulator = pkgs:
+          if builtins.elem final.system pkgs.qemu-user-static.passthru.qemuUserPlatforms then
```
Is there any reason to check and return null, rather than just always returning qemu-user-static and letting the meta checks produce an error if its an unsupported platform?

zhaofengli replied:
This is to check whether the emulated platform is supported instead of the host platform. Could you elaborate on what the change would look like?

alyssais replied:
The change would be to have `emulator` and `staticEmulator` have a consistent interface.  So like `emulator`, `staticEmulator` would throw if unsupported, and there'd be a `staticEmulatorAvailable` to mirror `emulatorAvailable`.  I just want to make sure we don't have two interfaces that work differently for such similar concepts.

alyssais commented on pkgs/applications/virtualization/qemu/default.nix:
```diff
@@ -178,7 +186,12 @@ stdenv.mkDerivation rec {
+    ++ lib.optional (!enableBlobs) "--disable-install-blobs"
+
+    # FIXME: "multiple definition of `strtoll'" with libnbcompat
+    ++ lib.optional stdenv.hostPlatform.isStatic "--static --extra-ldflags=-Wl,--allow-multiple-definition";
```
Would be nice if we could figure out why this is happening.  One definition is from libnbcompat — where's the other from?

zhaofengli replied:
Seems to be a conflict between libnbcompat and musl:

```
/nix/store/kbhvb5y763zm2xskz94j176qv2dvbv58-x86_64-unknown-linux-musl-binutils-2.40/bin/x86_64-unknown-linux-musl-ld: /nix/store/g29g3hydcrrjkisrmk96gnx23mnnldrj-musl-static-x86_64-unknown-linux-musl-1.2.3/li
b/libc.a(strtol.lo): in function `strtoll.localalias':                                                  
(.text.strtoll+0x0): multiple definition of `strtoll'; /nix/store/dp31apbsw47a9bm37w45kh96vkngh2az-compat-netbsd-static-x86_64-unknown-linux-musl-9.2/lib/libnbcompat.a(strtoll.o):(.text+0x0): first defined he
re
```

Also it appears that there are many more build dependencies that we can remove. I'll play around and see how much further we can reduce the input closure.

alyssais replied:
Thanks.  I wonder why we don't see this on other packages.  I'd like to take a look, but don't know when I'll have the capacity…

oxalica commented:
> When building coreutils for aarch64-linux on x86_64-linux with it, all tests pass except for `test-free` where it segfaults the tests become stuck

It is an memory allocation failure inside QEMU. I do not think it is an issue of packaging or musl-libc.

The test in located at `gnulib-tests/test-free.c`. It intentionally spams `mmap` until failure, in order to test a behavior under `munmap`-failure situation, and simply leaks them. The test returns from `main` successfully, but makes QEMU fail an allocation later when caching the JIT-ted code. If manually run the test, you can see a prompt like `riscv64-linux: GLib: ../glib/gmem.c:136: failed to allocate 40 bytes` before freezing.

It is [also intentional](https://bugzilla.gnome.org/show_bug.cgi?id=674446) that allocation failure in glib causes a infinite `for(;;);`, though in docs it says "the program is terminated" in case of OOM. The allocation failure point is [here](https://gitlab.gnome.org/GNOME/glib/-/blob/2.76.3/glib/gmem.c?ref_type=tags#L168), and the `g_error` definition is [here](https://gitlab.gnome.org/GNOME/glib/-/blob/2.76.3/glib/gmessages.h?ref_type=tags#L317).

<details><summary>Backtrace</summary>

```
#0  0x000000000057bf92 in g_malloc ()
#1  0x0000000000594fca in g_tree_insert_internal ()
#2  0x000000000049d800 in tcg_tb_insert ()
#3  0x00000000004e87a2 in tb_gen_code ()
#4  0x00000000004e2eee in cpu_exec_loop.constprop ()
#5  0x00000000004e3125 in cpu_exec_setjmp.isra ()
#6  0x00000000004e3688 in cpu_exec ()
#7  0x00000000004094a0 in cpu_loop ()
#8  0x0000000000401886 in main ()
```

</details>

ImBearChild commented:
This is a friendly ping. Since the allocation failure is intentional, would it be possible to merge the pull request now? I believe that this feature would be very useful for developers who work on embedded Linux products.

Detegr commented:
I ran into this limitation today once again. +1 for merging this.

muellerbernd commented:
I also ran into this problem. +1 for merging

ck3d commented:
Could you please solve the merge conflicts. I will merge afterwards.

TECHNOFAB11 commented:
Ran into this the last few days and took me a while to figure out why binfmt was working on Debian but not on NixOS, +1 for merging, would allow me to run different arch images in Kubernetes

bltavares commented:
I'm not familiar with the etiquette on updating the PR of others, but I did attempt to update it to fix merge conflicts and new compilation issues on https://github.com/NixOS/nixpkgs/pull/300070

I did not find yet the right git commands to keep https://github.com/NixOS/nixpkgs/pull/160802 authorship during the merge I'd like to thank @zhaofengli for the original PR and I'll investigate how to keep the authorship intact meanwhile

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/docker-ignoring-platform-when-run-in-nixos/21120/17

SuperSandro2000 commented:
@bltavares see https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

zhaofengli commented:
Superseded by #160802 (and previously #300070).

oxalica commented:
> Superseded by #160802 (and previously #300070).

Do you mean #334859? Your link is a self-reference to this one.

zhaofengli commented:
> Do you mean https://github.com/NixOS/nixpkgs/pull/334859? Your link is a self-reference to this one.

Oops, you’re right. I fixed the link.
